iris-mpc-cpu/src/hnsw/graph/layered_graph.rs

```diff
@@ -0,0 +1,362 @@
+// PlaintextStore to a secret shared VectorStore.
+impl<V: VectorStore> GraphMem<V> {
+    pub fn from_another<U, F1, F2>(graph: GraphMem<U>, vector_map: F1, distance_map: F2) -> Self
+    where
+        U: VectorStore,
+        F1: Fn(U::VectorRef) -> V::VectorRef + Copy,
+        F2: Fn(U::DistanceRef) -> V::DistanceRef + Copy,
+    {
+        let new_entry_point = graph.entry_point.map(|ep| EntryPoint {
+            point: vector_map(ep.point),
+            layer: ep.layer,
+        });
+
+        let new_layers: Vec<_> = graph
+            .layers
+            .into_iter()
+            .map(|v| {
+                let links: HashMap<_, _> = v
+                    .links
+                    .into_iter()
+                    .map(|(v, q)| (vector_map(v), q.map::<V, F1, F2>(vector_map, distance_map)))
+                    .collect();
+                Layer::<V> { links }
+            })
+            .collect();
+
+        GraphMem::<V> {
+            entry_point: new_entry_point,
+            layers:      new_layers,
+        }
+    }
+
+    pub async fn get_entry_point(&self) -> Option<(V::VectorRef, usize)> {
+        self.entry_point
+            .as_ref()
+            .map(|ep| (ep.point.clone(), ep.layer))
+    }
+
+    pub async fn set_entry_point(&mut self, point: V::VectorRef, layer: usize) {
+        if let Some(previous) = self.entry_point.as_ref() {
+            assert!(
+                previous.layer < layer,
+                "A new entry point should be on a higher layer than before."
+            );
+        }
+
+        for _ in self.layers.len()..=layer {
+            self.layers.push(Layer::new());
+        }
+
+        self.entry_point = Some(EntryPoint { point, layer });
+    }
+
+    pub async fn get_links(
+        &self,
+        base: &<V as VectorStore>::VectorRef,
+        lc: usize,
+    ) -> SortedNeighborhood<V> {
+        let layer = &self.layers[lc];
+        if let Some(links) = layer.get_links(base) {
+            links.clone()
+        } else {
+            SortedNeighborhood::new()
+        }
+    }
+
+    /// Set the neighbors of vertex `base` at layer `lc` to `links`.  Requires
+    /// that the graph already has been extended to have layer `lc` using the
+    /// `set_entry_point` function for an entry point at at least this layer.
+    ///
+    /// Panics if `lc` is higher than the maximum initialized layer.
+    pub async fn set_links(&mut self, base: V::VectorRef, links: SortedNeighborhood<V>, lc: usize) {
+        let layer = self.layers.get_mut(lc).unwrap();
+        layer.set_links(base, links);
+    }
+
+    pub async fn num_layers(&self) -> usize {
+        self.layers.len()
+    }
+
+    pub async fn connect_bidir(
+        &mut self,
+        vector_store: &mut V,
+        q: &V::VectorRef,
+        neighbors: SortedNeighborhood<V>,
+        max_links: usize,
+        lc: usize,
+    ) {
+        // let M = self.params.get_M(lc);
```

Comment:
remove?

Reply:
or just merge `main` due to [PR 963](https://github.com/worldcoin/iris-mpc/pull/963)